javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -73,71 +76,91 @@ private static int treeSize(int branchCount, int depthShift) {
+            array = result.modifyLeaf(array, shift, offset - 1, COPY_NODE, prependToLeaf(iterator, delta));
+            result = new BitMappedTrie<>(array, offset - delta.val, result.length + delta.val, shift);
         }
+        return result;
     }
     private boolean isFullLeft() { return offset == 0; }
+    private NodeModifier prependToLeaf(java.util.Iterator<? extends T> iterator, MutableInt delta) {
+        return (array, index) -> {
+            final Object[] copy = copy(array, BRANCHING_FACTOR);
+            for (; iterator.hasNext() && index >= 0; index--) {
+                copy[index] = iterator.next();
+                delta.val++;
+            }
+            return copy;
+        };
+    }
 
-    BitMappedTrie<T> append(T trailing) {
-        final int newSize = length() + 1;
-        if (length() == 0) {
-            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
-        } else {
-            Object[] array = this.array;
-            int shift = depthShift;
-            if (isFullRight(newSize)) {
+    BitMappedTrie<T> appendAll(Iterable<? extends T> iterable) {
+        BitMappedTrie<T> result = this;
+        for (java.util.Iterator<? extends T> iterator = iterable.iterator(); iterator.hasNext(); ) {
+            Object[] array = result.array;
+            int shift = result.depthShift;
+            if (result.isFullRight()) {
                 array = asArray(array);
                 shift += BRANCHING_BASE;
             }
 
-            array = modifyLeaf(array, shift, offset + length(), COPY_NODE, updateLeafWith(trailing));
-            return new BitMappedTrie<>(array, offset, newSize, shift);
+            final MutableInt delta = new MutableInt(0);
+            array = result.modifyLeaf(array, shift, result.offset + result.length, COPY_NODE, appendToLeaf(iterator, delta));
+            result = new BitMappedTrie<>(array, result.offset, result.length + delta.val, shift);
         }
+        return result;
+    }
+    private boolean isFullRight() { return (offset + length + 1) > treeSize(BRANCHING_FACTOR, depthShift); }
+    private NodeModifier appendToLeaf(java.util.Iterator<? extends T> iterator, MutableInt delta) {
+        return (array, index) -> {
+            final Object[] copy = copy(array, BRANCHING_FACTOR);
+            for (; iterator.hasNext() && index < copy.length; index++) {
+                copy[index] = iterator.next();
+                delta.val++;
+            }
+            return copy;
+        };
     }
-    private boolean isFullRight(int newSize) { return (offset + newSize) > treeSize(BRANCHING_FACTOR, depthShift); }
 
     BitMappedTrie<T> update(int index, T element) {
         final Object[] root = modifyLeaf(array, depthShift, offset + index, COPY_NODE, updateLeafWith(element));
-        return new BitMappedTrie<>(root, offset, length(), depthShift);
+        return new BitMappedTrie<>(root, offset, length, depthShift);
```

Comment:
@danieldietrich requested previously that I call `length` via the field, not the method.
Now this class is encapsulated enough that I applied it :)